src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ScheduleService.kt

```diff
@@ -42,51 +39,34 @@ class ScheduleService(
 
   private fun CourseScheduleRequest.getScheduleStatus() = if (cancelled) "CANC" else "SCH"
 
-  fun buildNewSchedules(scheduleRequests: List<CourseScheduleRequest>, courseActivity: CourseActivity): List<CourseSchedule> =
-    try {
-      mapSchedules(scheduleRequests, courseActivity)
-        .let { requestedSchedules ->
-          findPastSchedules(requestedSchedules, courseActivity) + findOrAddFutureSchedules(
-            requestedSchedules,
-            courseActivity,
-          )
-        }
-    } catch (ex: Exception) {
-      log.error("Failed to build new requested schedules $scheduleRequests for existing schedules ${courseActivity.courseSchedules} due to error", ex)
-      throw ex
+  /*
+   * Build a list of schedules to save against the Activity:
+   * - any past schedules that are already saved are considered to be immutable, you can't change history
+   * - any attempts to change immutable schedules are ignored (which means the request doesn't need to include the full schedule history, we won't delete missing old schedules)
+   * - updatable schedules are any saved schedules from today onwards included in the request
+   * - schedules from today onwards not included in the request are deleted
+   * - unsaved schedules are any schedules included in the request but not yet saved
+   */
```

Comment:
This is the main change. The result is that we've simplified the update of multiple schedules while also making it more flexible.